src/Generator.cpp

```diff
@@ -0,0 +1,53 @@
+#include "Generator.h"
+
+Generator::Generator() {}
+
+void Generator::initialize_deques(std::vector<I32> x, std::vector<I32> y)
+{
+	for(I32 element : x)
+	{
+		x_coordinates.push_back(element);
+	}
+	for(I32 element : y)
+	{
+		y_coordinates.push_back(element);
+	}
+	std::random_shuffle(x_coordinates.begin(), x_coordinates.end());
+	x_coordinates.push_back(-1);
+	std::random_shuffle(y_coordinates.begin(), y_coordinates.end());
+	y_coordinates.push_back(-1);
+}
+
+I32 Generator::next_x()
+{
+	I32 next_element = x_coordinates.front();
+	x_coordinates.pop_front();
+	if (next_element == -1)
+	{
+		std::random_shuffle(x_coordinates.begin(), x_coordinates.end());
+		x_coordinates.push_back(-1);
+		return this->next_x();
+	}
+	else
+	{
+		x_coordinates.push_back(next_element);
+		return next_element;
+	}
+}
+
+I32 Generator::next_y()
+{
+	I32 next_element = y_coordinates.front();
+	y_coordinates.pop_front();
+	if (next_element == -1)
+	{
+		std::random_shuffle(y_coordinates.begin(), y_coordinates.end());
+		y_coordinates.push_back(-1);
```

Comment:
I think there's a bug here, but try it out with a small list of 3 elements. I think you're going to lose elements along the way, one every time...

Reply:
Hmm. I tried printing out the deque every iteration and reducing the time between iterations to test it better, but I can't seem to replicate that bug. No elements are being lost on my end.